Congratulations! After a lot of hard work in the data munging mines, you've landed a job as Lead Analyst for an independent gaming company. You've been assigned the task of analyzing the data for their most recent fantasy game Heroes of Pymoli.

Like many others in its genre, the game is free-to-play, but players are encouraged to purchase optional items that enhance their playing experience. As a first task, the company would like you to generate a report that breaks down the game's purchasing data into meaningful insights.


In [3]:
#Dependencies
import pandas as pd
import numpy as np

In [4]:
#Save path to data set in a variable
data_file = "purchase_data.json"
data_file2 = "purchase_data2.json"


In [5]:
#Use pandas to read the data
print(data_file)
#data_file_pd = pd.read_json(data_file)
pd_file = pd.read_json(data_file)
pd_file.head()

purchase_data.json


Age Gender  Item ID                                  Item Name  Price  \
0   38   Male      165                Bone Crushing Silver Skewer   3.37   
1   21   Male      119  Stormbringer, Dark Blade of Ending Misery   2.32   
2   34   Male      174                            Primitive Blade   2.46   
3   21   Male       92                               Final Critic   1.36   
4   23   Male       63                             Stormfury Mace   1.27   

             SN  
0     Aelalis34  
1        Eolo46  
2   Assastnya25  
3  Pheusrical25  
4        Aela59

Purchasing Analysis (Total)

In [6]:
players_count = len(pd_file["SN"].value_counts())
players_count

573

In [7]:
#Unique Number of Items
unique_purchases = len(pd_file["Price"].unique())
unique_purchases


152

In [8]:
price_mean = pd_file["Price"].mean()
price_mean

2.931192307692303

In [9]:
#Total Number of Prurchases
purch_count = len(pd_file["Price"])
purch_count

780

In [10]:
totalRev = sum(pd_file["Price"])
totalRev

2286.3299999999963

In [11]:
player_count = pd.DataFrame({"Player Count": players_count},
                           index = [0])
player_count

Player Count
0           573

In [12]:
#Convert series into DataFrams
purchaseAnalysis= pd.DataFrame({"Number of Unique Items": unique_purchases,
                           "Average Purchase Price": price_mean,
                           "Total Number of Purchases": purch_count,
                           "Total Revenue": totalRev},
                          index = [0])
purchaseAnalysis

Average Purchase Price  Number of Unique Items  Total Number of Purchases  \
0                2.931192                     152                        780   

   Total Revenue  
0        2286.33

You must include a written description of three observable trends based on the data.

In the above dataframe, I see that the average price is 2.93 overall. The average purchase price for female was $2.81, $2.95 for men and $3.24 for other/non-disclosed. It appears that as a whole, women paid the least for the game. However, other/non-disclosed were willing to pay the most.

Unsurprisingly, men bought the most video games by far! Out of 780 purchases, men bought 633. That is roughly 81%! 

The total purchase value is also very high for men: $1,867.68. By contrast, the purchase value was only 382.91 for women. That is roughly a 79% difference. 

Overall, the independent gaming company should focus on marketing their products to men if they already aren't doing so to get the most sales and profit. Women only make up roughly 17% of purchases. 

Additionally, the most profitable item was Retribution Axe for 9 years old. The independent gaming company may also want to focus on selling this product in as many stores as possible.

Furthermore, the greatest purchase value was experience for 20 to 24 year olds. The independent gaming company may benefit from selling games in colleges, as a lot of students in college are in that age range.

# Gender Demographics

In [13]:
#Gender Demographics
#Percentage and Counts of Male/Female Players
#Percentage and Counts of Other/Non-Disclosed
#same player can have multiple items 
male_count_1 = pd_file.groupby(["SN", "Gender"]).size()
#male_count_1.count().head()

players_1 = pd.DataFrame(male_count_1)
players_1
players_2 = pd.DataFrame(players_1.groupby(["Gender"]).count())
players_2.columns = ["GenderCount"]
players_2


GenderCount
Gender                            
Female                         100
Male                           465
Other / Non-Disclosed            8

In [14]:
players_male = players_2["GenderCount"].loc["Male"]
print(players_male)

465


In [15]:
male_percentage = round(players_male / player_count, 2) * 100
male_percentage
#male_percentage * 100

Player Count
0          81.0

In [16]:
# female_players = len(pd_file.loc[pd_file["Gender"] == "Female"])
# female_players
female_players = players_2["GenderCount"].loc["Female"]
female_players

100

In [17]:
female_percentage = round(female_players / player_count, 2) * 100
female_percentage
#female_percentage * 100

Player Count
0          17.0

In [18]:
# other_players = len(pd_file[pd_file["Gender"] == "Other / Non-Disclosed"])
# other_players
other_players = players_2["GenderCount"].loc["Other / Non-Disclosed"]
other_players

8

In [19]:
other_percentage = round(other_players / player_count, 2) * 100
other_percentage 
#other_percentage * 100

Player Count
0           1.0

In [20]:
#Convert series into DataFrames
genderDemo= pd.DataFrame({"Count of Male Players": players_male,
                           "Percent of Male Players": male_percentage,
                           "Count of Female Players": female_players,
                           "Percent of Female Players": female_percentage,
                           "Count of Other / Non-Disclosed": other_players,
                           "Percent of Other / Non-Disclosed": other_percentage},
        
                          index = [0])
genderDemo

Count of Female Players  Count of Male Players  \
0                      100                    465   

   Count of Other / Non-Disclosed Percent of Female Players  \
0                               8                        17   

  Percent of Male Players Percent of Other / Non-Disclosed  
0                      81                                1

# Purchasing Analysis (Gender)

In [21]:
#Gender Purchase Analysis
purchGender = pd_file.groupby(["Gender"])
purchGender.head()

Age                 Gender  Item ID  \
0     38                   Male      165   
1     21                   Male      119   
2     34                   Male      174   
3     21                   Male       92   
4     23                   Male       63   
7     29                 Female      169   
16    22                 Female      123   
17    22                 Female       59   
22    11                 Female       11   
29    16                 Female       45   
177   34  Other / Non-Disclosed      155   
209   33  Other / Non-Disclosed      157   
244   21  Other / Non-Disclosed      183   
267   33  Other / Non-Disclosed       65   
276   12  Other / Non-Disclosed      128   

                                     Item Name  Price              SN  
0                  Bone Crushing Silver Skewer   3.37       Aelalis34  
1    Stormbringer, Dark Blade of Ending Misery   2.32          Eolo46  
2                              Primitive Blade   2.46     Assastnya25  
3                                 Final Critic   1.36    Pheusrical25  
4                               Stormfury Mace   1.27          Aela59  
7       Interrogator, Blood Blade of the Queen   3.32    Iathenudil29  
16                           Twilight's Carver   1.14      Sundista85  
17               Lightning, Etcher of the King   1.65       Aenarap34  
22                                   Brimstone   2.52        Deural48  
29                         Glinting Glass Edge   2.46       Phaedai25  
177                  War-Forged Gold Deflector   3.73       Assassa38  
209                    Spada, Etcher of Hatred   2.21  Frichistasta59  
244                        Dragon's Greatsword   2.36      Tyaerith73  
267                  Conqueror Adamantite Mace   1.96  Frichistasta59  
276              Blazeguard, Reach of Eternity   4.00      Aillycal84

In [22]:
gnpurcount = pd_file["Gender"].value_counts()
#gnpurcount = pd.DataFrame(purchGender[3].value_counts())
gnpurcount

Male                     633
Female                   136
Other / Non-Disclosed     11
Name: Gender, dtype: int64

In [23]:
groupprice1 = purchGender["Price"].mean()
groupprice1

Gender
Female                   2.815515
Male                     2.950521
Other / Non-Disclosed    3.249091
Name: Price, dtype: float64

In [24]:
purchbygen = purchGender["Price"].sum()
purchbygen

Gender
Female                    382.91
Male                     1867.68
Other / Non-Disclosed      35.74
Name: Price, dtype: float64

In [25]:
normtotal = round(purchbygen/players_2["GenderCount"],2)
normtotal

Gender
Female                   3.83
Male                     4.02
Other / Non-Disclosed    4.47
dtype: float64

In [26]:
genderpa = pd.DataFrame({"Purchase Count":gnpurcount,
                        "Average Purchase Price":groupprice1,
                         "Total Purchase Value":purchbygen,
                          "Normalized Total": normtotal} )
genderpa

Average Purchase Price  Normalized Total  \
Female                               2.815515              3.83   
Male                                 2.950521              4.02   
Other / Non-Disclosed                3.249091              4.47   

                       Purchase Count  Total Purchase Value  
Female                            136                382.91  
Male                              633               1867.68  
Other / Non-Disclosed              11                 35.74

# AGE DEMOGRAPHICS

In [27]:
# Figuring out the maximum age can help with creating buckets. 
print(pd_file["Age"].max())

45


In [28]:
#Age Demographics
#Break into Bins of 4 Years
bins = [0, 9.9, 14.9, 19.9, 24.9, 29.9, 34.9, 39.9, 120]
group_names = ["<10", "10-14", "15-19", "20-24", "25-29",
             "30-34", "35-39", "40+"]
group_names

['<10', '10-14', '15-19', '20-24', '25-29', '30-34', '35-39', '40+']

In [29]:
pd.cut(pd_file["Age"], bins, labels = group_names).head()

0    35-39
1    20-24
2    30-34
3    20-24
4    20-24
Name: Age, dtype: category
Categories (8, object): [<10 < 10-14 < 15-19 < 20-24 < 25-29 < 30-34 < 35-39 < 40+]

In [79]:
# pymoli_df["Age Group"] = pd.cut(pymoli_df["Age"], bins, labels=group_names)
pd.cut(pd_file["Age"], bins, labels = group_names)
pd_file["Age Range"] = pd.cut(pd_file["Age"], bins, labels = group_names)
pdages = pd_file.groupby(["SN","Age Range"])

                                                              
players = pd.DataFrame(pdages.size())
#players

agegroups = pd.DataFrame(players.groupby(["Age Range"]).count())
agegroups.columns= ["Total Count"]
agegroups["Percentage of Players"] = round(100 * agegroups["Total Count"]/ agegroups["Total Count"].sum(),2)
agegroups

Total Count  Percentage of Players
Age Range                                    
<10                 19                   3.32
10-14               23                   4.01
15-19              100                  17.45
20-24              259                  45.20
25-29               87                  15.18
30-34               47                   8.20
35-39               27                   4.71
40+                 11                   1.92

In [80]:
pd_file["Age Range"] = pd.cut(pd_file["Age"], bins, labels = group_names)
pd_file.head()

Age Gender  Item ID                                  Item Name  Price  \
0   38   Male      165                Bone Crushing Silver Skewer   3.37   
1   21   Male      119  Stormbringer, Dark Blade of Ending Misery   2.32   
2   34   Male      174                            Primitive Blade   2.46   
3   21   Male       92                               Final Critic   1.36   
4   23   Male       63                             Stormfury Mace   1.27   

             SN Age Range  
0     Aelalis34     35-39  
1        Eolo46     20-24  
2   Assastnya25     30-34  
3  Pheusrical25     20-24  
4        Aela59     20-24

In [81]:
# percentplayers = round(purchcount / players_count,2) * 100

# percentplayers

# Purchasing Analysis (Age)

In [82]:
#Purchase Count
#Analyze the data.
#Groupby demographics
demoGroup = pd_file.groupby("Age Range")
#How many rows per bin
print(demoGroup["Age Range"].count())
#Average of each column within GroupBy
demoGroup.mean()

Age Range
<10       28
10-14     35
15-19    133
20-24    336
25-29    125
30-34     64
35-39     42
40+       17
Name: Age Range, dtype: int64


Age     Item ID     Price
Age Range                                 
<10         7.535714   82.321429  2.980714
10-14      12.171429   90.600000  2.770000
15-19      16.631579   94.443609  2.905414
20-24      21.875000   86.038690  2.913006
25-29      26.200000   93.832000  2.962640
30-34      31.609375  101.453125  3.082031
35-39      36.714286  107.071429  2.842857
40+        40.588235   90.823529  3.161765

In [83]:
purchcount = demoGroup["Age Range"].count()
purchcount

Age Range
<10       28
10-14     35
15-19    133
20-24    336
25-29    125
30-34     64
35-39     42
40+       17
Name: Age Range, dtype: int64

In [84]:
#Groupby demographics
#demoGroup = pd_file.groupby("Age Range")
#How many rows per bin
agegroupprice = round(demoGroup["Price"].mean(),2)
agegroupprice

Age Range
<10      2.98
10-14    2.77
15-19    2.91
20-24    2.91
25-29    2.96
30-34    3.08
35-39    2.84
40+      3.16
Name: Price, dtype: float64

In [85]:
purchbyage = round(demoGroup["Price"].sum(),2)
purchbyage

Age Range
<10       83.46
10-14     96.95
15-19    386.42
20-24    978.77
25-29    370.33
30-34    197.25
35-39    119.40
40+       53.75
Name: Price, dtype: float64

In [86]:
normage = round(purchbyage/purchcount,2)
normage

Age Range
<10      2.98
10-14    2.77
15-19    2.91
20-24    2.91
25-29    2.96
30-34    3.08
35-39    2.84
40+      3.16
dtype: float64

In [87]:
#Create a dataframe with Age, Purchase Count, Average Purchase Price, Total Purchase Value and Normalized Totals

ageanalysis = pd.DataFrame({"Purchase Count":purchcount,
                        "Average Purchase Price":agegroupprice,
                         "Total Purchase Value":purchbyage,
                          "Normalized Total": normage} )
ageanalysis

Average Purchase Price  Normalized Total  Purchase Count  \
Age Range                                                             
<10                          2.98              2.98              28   
10-14                        2.77              2.77              35   
15-19                        2.91              2.91             133   
20-24                        2.91              2.91             336   
25-29                        2.96              2.96             125   
30-34                        3.08              3.08              64   
35-39                        2.84              2.84              42   
40+                          3.16              3.16              17   

           Total Purchase Value  
Age Range                        
<10                       83.46  
10-14                     96.95  
15-19                    386.42  
20-24                    978.77  
25-29                    370.33  
30-34                    197.25  
35-39                    119.40  
40+                       53.75

Top Spenders

In [88]:
#Top 5 spenders by total purchase value
topspender = pd_file.groupby("SN").sum()
topspender = topspender.sort_values('Price').tail()
del topspender["Item ID"]
del topspender["Age"]
topspend = topspender.sort_values('Price').tail()
print(topspend)

             Price
SN                
Eoda93       11.58
Haellysu29   12.73
Mindimnya67  12.74
Saedue76     13.56
Undirrala66  17.06


In [89]:
#Top 5 spenders total purchases
numpurch = pd.DataFrame(pd_file.groupby("SN").count())

#Top 5 spenders total purchase value
topspender = pd.DataFrame(pd_file.groupby("SN").sum())

#Sort the total purchase price. The default to sort_values is ascending. If I put , ascending=False would be descending.
topspender = topspender.sort_values(by=["Price"])


topspender["Purchase Count"] = numpurch["Item ID"]
topspender["Average Purchase Price"] = round(topspender["Price"]/topspender["Purchase Count"],2).map("${:,.2f}".format)
topspender["Total Purchase Value"] = topspender["Price"].map("${:,.2f}".format)

#I could have done topspender.head() if I sorted descending. 
topspender.tail()

Age  Item ID  Price  Purchase Count Average Purchase Price  \
SN                                                                        
Eoda93        66      284  11.58               3                  $3.86   
Haellysu29    63      353  12.73               3                  $4.24   
Mindimnya67  156      609  12.74               4                  $3.18   
Saedue76     100      233  13.56               4                  $3.39   
Undirrala66  145      472  17.06               5                  $3.41   

            Total Purchase Value  
SN                                
Eoda93                    $11.58  
Haellysu29                $12.73  
Mindimnya67               $12.74  
Saedue76                  $13.56  
Undirrala66               $17.06

In [90]:
#pd_file.head()
pd_file = pd_file.drop(labels = ["Age Range"],axis = 1)
pd_file.head()

Age Gender  Item ID                                  Item Name  Price  \
0   38   Male      165                Bone Crushing Silver Skewer   3.37   
1   21   Male      119  Stormbringer, Dark Blade of Ending Misery   2.32   
2   34   Male      174                            Primitive Blade   2.46   
3   21   Male       92                               Final Critic   1.36   
4   23   Male       63                             Stormfury Mace   1.27   

             SN  
0     Aelalis34  
1        Eolo46  
2   Assastnya25  
3  Pheusrical25  
4        Aela59

In [91]:
topspendpurchcount = pd.DataFrame(pd_file.groupby(["Item ID", "Item Name"]).count())
topspendpurchcount.head()

Age  Gender  Price  SN
Item ID Item Name                                 
0       Splinter              1       1      1   1
1       Crucifer              4       4      4   4
2       Verdict               1       1      1   1
3       Phantomlight          1       1      1   1
4       Bloodlord's Fetish    1       1      1   1

# Most Popular Items

In [92]:
#Most Popular Items
topspendpurchcount = pd.DataFrame(pd_file.groupby(["Item ID", "Item Name"]).count())
#print(topspendpurchcount)

#Top 5 spenders total purchase value
topitemvalue = pd.DataFrame(pd_file.groupby(["Item ID","Item Name"]).sum()["Price"])
#topitemvalue

#Renaming SN to Purchase Count as it will show how many times the item was purchased. (A person purchases an item.)
topspendpurchcount["Purchase Count"] = topspendpurchcount["SN"]

#Item Price is the sum of prices paid per item divided by the total number of people who bought it.
topspendpurchcount["Item Price"] = round(topitemvalue["Price"]/topspendpurchcount["SN"],2)

#Total Purchase Value is the sum of prices. This is in the topvalueitem dataframe.
topspendpurchcount["Total Purchase Value"] = topitemvalue["Price"]

#The most popular items by purchase count would require us to sort by purchase count descending (highest to lowest).
topspendpurchcount = topspendpurchcount.sort_values(by = ["Purchase Count","Total Purchase Value"],ascending = False)
#topspendpurchcount = topspendpurchcount.sort_values(by = ["Purchase Count"],ascending = False)

#Deleting columns that are not necessary to output. This was not deleted from our original data source, pd_file.
topspendpurchcount = topspendpurchcount.drop(labels = ["Age","Gender","Price","SN"],axis = 1)

##Format the columns for currency as they are monetary values.
topspendpurchcount["Item Price"] = topspendpurchcount["Item Price"].map("${:,.2f}".format)
topspendpurchcount["Total Purchase Value"] = topspendpurchcount["Total Purchase Value"].map("${:,.2f}".format)

#Top 5 Popular Items
topspendpurchcount.head(5)


Purchase Count Item Price  \
Item ID Item Name                                                         
39      Betrayal, Whisper of Grieving Widows              11      $2.35   
84      Arcane Gem                                        11      $2.23   
34      Retribution Axe                                    9      $4.14   
31      Trickster                                          9      $2.07   
13      Serenity                                           9      $1.49   

                                             Total Purchase Value  
Item ID Item Name                                                  
39      Betrayal, Whisper of Grieving Widows               $25.85  
84      Arcane Gem                                         $24.53  
34      Retribution Axe                                    $37.26  
31      Trickster                                          $18.63  
13      Serenity                                           $13.41

Most Profitable Items

In [93]:
#Most Profitable Items
topspendpurchcount = pd.DataFrame(pd_file.groupby(["Item ID", "Item Name"]).count())
#print(topspendpurchcount)

#Top 5 spenders total purchase value
topitemvalue = pd.DataFrame(pd_file.groupby(["Item ID","Item Name"]).sum()["Price"])
#topitemvalue

#Renaming SN to Purchase Count as it will show how many times the item was purchased. (A person purchases an item.)
topspendpurchcount["Purchase Count"] = topspendpurchcount["SN"]

#Item Price is the sum of prices paid per item divided by the total number of people who bought it.
topspendpurchcount["Item Price"] = round(topitemvalue["Price"]/topspendpurchcount["SN"],2)

#Total Purchase Value is the sum of prices. This is in the topvalueitem dataframe.
topspendpurchcount["Total Purchase Value"] = topitemvalue["Price"]

#The most popular items by purchase count would require us to sort by purchase count descending (highest to lowest).
topspendpurchcount = topspendpurchcount.sort_values(by = ["Total Purchase Value", "Purchase Count"],ascending = False)
#topspendpurchcount = topspendpurchcount.sort_values(by = ["Purchase Count"],ascending = False)

#Deleting columns that are not necessary to output. This was not deleted from our original data source, pd_file.
topspendpurchcount = topspendpurchcount.drop(labels = ["Age","Gender","Price","SN"],axis = 1)

##Format the columns for currency as they are monetary values.
topspendpurchcount["Item Price"] = topspendpurchcount["Item Price"].map("${:,.2f}".format)
topspendpurchcount["Total Purchase Value"] = topspendpurchcount["Total Purchase Value"].map("${:,.2f}".format)

#Top 5 Popular Items
topspendpurchcount.head(5)

Purchase Count Item Price  \
Item ID Item Name                                               
34      Retribution Axe                          9      $4.14   
115     Spectral Diamond Doomblade               7      $4.25   
32      Orenmir                                  6      $4.95   
103     Singed Scalpel                           6      $4.87   
107     Splitter, Foe Of Subtlety                8      $3.61   

                                   Total Purchase Value  
Item ID Item Name                                        
34      Retribution Axe                          $37.26  
115     Spectral Diamond Doomblade               $29.75  
32      Orenmir                                  $29.70  
103     Singed Scalpel                           $29.22  
107     Splitter, Foe Of Subtlety                $28.88

In [94]:
most_prof_total = pd.DataFrame(pd_file.groupby(['Item ID','Item Name']).count())
most_prof_total.head()
most_prof_total = most_prof_total.groupby(["Item ID","Item Name"]).sum()["Price"]
most_prof_total.head()

Item ID  Item Name         
0        Splinter              1
1        Crucifer              4
2        Verdict               1
3        Phantomlight          1
4        Bloodlord's Fetish    1
Name: Price, dtype: int64

In [95]:
 #Purchase Count
topspendpurchcount = pd_file.groupby("SN").count()
topspendpurchcount = topspendpurchcount.drop(["Age","Gender","Item Name","Price"], axis = 1)
spendpurch = topspendpurchcount.sort_values("Item ID").tail()
spendpurch = spendpurch.rename(columns = {'Item ID': 'Purchase Count'})
print(spendpurch)

             Purchase Count
SN                         
Sondastan54               4
Qarwen67                  4
Mindimnya67               4
Hailaphos89               4
Undirrala66               5


In [96]:
#Dataframe by age
#Gender Purchase Analysis
age_df = pd_file.groupby(["Age"])
#age_df.head()

In [97]:
#Purchase Count
agepurcount = pd_file["Age"].value_counts()
agepurcount.head()

20    98
24    70
22    68
25    67
23    57
Name: Age, dtype: int64

In [98]:
#Total Purchase Value
purchbyage = age_df["Price"].sum()
#purchbyage